### PR TITLE
Implement SubType support (overriding class)

### PR DESCRIPTION
**Pull request by romaninsh**
Implementation of SubType substitution: #340. Simple use:

``` php
get_class($account->ref('Transactions')->load(123))  // Transaction_Deposit
get_class($account->ref('Transactions')->load(124))  // Transaction_Transfer
```

### CODE REVIEW

**Comment by romaninsh** [comment]
@DarkSide666 added improved coverage..

File: docs/advanced.rst [hunk 1]
```diff
@@ -6,6 +6,118 @@ Advanced Topics
...
+substitution::
+
+    foreach($account->ref('Transactions', ['typeSubstitution'=>true]) as $tr) {
+
+        $tr->verify();  // verify() method can be overloaded!
+    }
+
+
+    // however, for export, we don't need expensive substitution
+    $stats = $account->ref('Transaction')->export();
```
Comment (DarkSide666) [inline docs/advanced.rst]: Better name $all_transactions not $stats

File: src/Util/DeepCopy.php [hunk 2]
```diff
@@ -79,19 +88,47 @@ public function to(Model $destination)
      */
     public function with(array $references)
     {
-        $this->references = array_merge_recursive($this->references, $references);
+        $this->references = $references;
+
+        return $this;
+    }
+
+    public function excluding(array $exclusions)
```
Comment (DarkSide666) [inline src/Util/DeepCopy.php]: not commented method

File: src/Util/DeepCopy.php [hunk 3]
```diff
@@ -79,19 +88,47 @@ public function to(Model $destination)
...
+     *
+     * @param $array
+     *
+     * @return array
+     */
+    protected function extractKeys($array): array
+    {
+        $result = [];
+        foreach ($array as $key=>$val) {
+            if (is_numeric($key)) {
```
Comment (DarkSide666) [inline src/Util/DeepCopy.php]: shouldn't we use `is_int` here? For example, key `'12.345'` is numeric, but most likely not what you expect here.

File: tests/SubTypesTest.php [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        if ($this->type) {
+            $this->addCondition('type', $this->type);
+        }
+        $this->addField('amount');
+
+        $this->addHook('afterLoad', function (STGenericTransaction $m) {
+            if (get_class($this) != $m->getClassName()) {
+                $cl = '\\'.$this->getClassName();
+                $cl = new $cl($this->persistence);
+                $cl->load($m->id);
```
Comment (DarkSide666) [inline tests/SubTypesTest.php]: executes SQL select one more time :(
bad for performance.

File: docs/advanced.rst [hunk 5]
```diff
@@ -6,6 +6,118 @@ Advanced Topics
...
+    }
+
+ATK Data allow class substitution during load and iteration by breaking "afterLoad"
+hook. Place the following inside Transaction::init()::
+
+    $this->addHook('afterLoad', function ($m) {
+        if (get_class($this) != $m->getClassName()) {
+            $cl = '\\'.$this->getClassName();
+            $cl = new $cl($this->persistence);
+            $cl->load($m->id);
```
[the remark this answers is not in the document]
Comment (romaninsh) [reply docs/advanced.rst]: no, it may contain extra fields or joins.

**Pull request merged**